### ISSUE TITLE
API: concurrency support / point-in-time consistency

### ISSUE THREAD

**Issue by bkmartinjr** [single-cell-data/TileDB-SOMA#540]
The current SOMA API lacks any support to access arrays/dataframes at a consistent state, and assumes that all objects are immutable for the purposes of reading.  For example, the current Python `SparseNdArray` "read" code contains (psuedo-code):
```
...
open array to get shape
close array
...
open array to get data
close array
...
```

A similar issue can arise when there are simultaneous writes and the API user performs more than one operation on an object over time, or has expectations of multi-object consistency (eg, for a SOMAExperiment).

Despite the underlying storage engine containing sophisticated features to support point-in-time consistency (and more) on _single_ objects, there is no way to access these features via the SOMA API.  I believe there are several options, in order of increasing complexity & utility:
1. leave as is - no API support. It is the responsibility of SOMA users to manage all concurrency and consistency.
2. for a single object (e.g., ndarray, dataframe), add the ability to read/write with a "recent" consistent state
3. \# 2, but with time-travel and roll-backs
4. Multi-object consistency (either "recent", aka \# 2 or with time-travel/roll-backs)

I propose we shoot for # 2 in the initial SOMA API release.  We can consider the others for future API versions, as needs dictate. The rest of this issue is a sketch of how the API might change.

CC: @thetorpedodog @johnkerl @aaronwolen @eddelbuettel @mlin @gspowley 

----

The current SOMA objects are stateless URI containers, which open the underlying storage objects upon each API call (sometimes more than once per API call).  Proposed API changes:
* each SOMA object (eg SOMASparseNdArray) becomes stateful, with two chunks of internal state:
    * an "open state", which is logically one of `not open`, `open for read`, `open for write`
    * a private handle to the underlying storage engine state require
* each SOMA object has two new methods - `open` and `close`, which put the object into one of the above states
* the read & write methods (and related methods such as `nnz`, `shape`, etc) all require that the object be in the correct mode for the operation.  Eg.,   somaNdArray.read() will raise an error if the object is not opened for reading.
* each SOMA implementation provides concurrency/consistency guarantees as able

The above should be sufficient to allow a TIileDB SOMA to guarantee that:
* objects that are opened for reading can perform multiple read-like operations at a given point of consistency
* objects that are opened for writing can treat open/close as barriers for noting eventual consistency
In other words, the same consistency contract made by the existing native TileDB-Py and TileDB-R API.

For example, with this new API, reading an dataframe would look like:
```
    uri = ...
    df = soma.IndexedDataFrame(uri)
    df.open(mode="read")
    some_data = df.read(...)
    df.close()
```

Operations classified by the mode they require:
* read mode:  required for `nnz`, `read*`, `shape`, `schema`, `soma_type`, metadata access, etc.
* write mode: required for `write*`, `reshape` (if we ever do it), etc.
* not-open mode:  required for `create`, `delete`, etc.

Side note: for languages that support it, ease of use features such as Python context manager support, would be very desirable.

**Comment by johnkerl** [comment]
Status from @mlin is we can close this now that #892 is merged -- pending a high-level check with @bkmartinjr  and @Shelnutt2  that it suffices for the overall goals

**Comment by mlin** [comment]
@bkmartinjr @Shelnutt2 Yea now that we've made all these code changes and put out an alpha release, just wanted to circle back and make sure we've met the high-level goal. To review,

1. `SOMATileDBContext` has slots for read and write timestamps (also for read timestamp *range*, although that's de-emphasized)
2. The read timestamp defaults to the moment when the `SOMATileDBContext` object is initialized.
3. When you open a collection, all accessed arrays and sub-collections inherit the `SOMATileDBContext` and therefore the timestamps. That's true whether you give it a context explicitly, or let it create its own default one.
4. Objects open in *write* mode ignore any read timestamp, and instead initialize themselves with the latest metadata/schema/elements/etc.
5. If the write timestamp is specified, all writes on objects using the context will share that timestamp, otherwise they'll each get their own individual 'now' timestamps.

Various alternate designs were possible, but I'm happy with how this turned out (@thetorpedodog having done the true heavy lifting, of course). In particular, setting the timestamps in `SOMATileDBContext` turned out very naturally since it provides inheritance from collection to accessed elements.

cc @johnkerl @atolopko-czi @gspowley 

Thanks all!

**Comment by bkmartinjr** [comment]
@mlin - with the large caveat that I have not yet had a chance to write significant code to this new capability, it seems both good and sufficient. Honestly, the only weakness I see is the lack of end-user docs/guidance providing best practice recipes for achieving specific goals, eg.,
* how to create an tree of collections/objects which a single consistent timestamp (eg, the census use case)
* on the same layered collection, how to make multiple modifications (eg. to an Experiment) saved in a way that is read-accessible at a single timestamp. And how to then read the multiple consistent checkpoints.

Most of the above boils down to simple patterns  -- eg., use a single context -- and some demo code, but still would be good to memorialize for end users.

TL;DR - I'm good resolving this GH issue.  @atolopko-czi is currently porting the census builder, so it would be good to hear how about his experiences.

**Comment by atolopko-czi** [comment]
Other than https://github.com/single-cell-data/TileDB-SOMA/issues/919, I've had no issues with the new open/close/shared context/PIT API changes. 

I believe this already planned, so just reiterating that some doc-level guidance on the semantics of on how to interleave reads & writes would be helpful. E.g. "use separate open() calls for explicit read and write handles", but then add some details about the various time-stamping strategies may affect visibility.

**Comment by bkmartinjr** [comment]
> doc-level guidance

Ideally including examples of common use cases

**Comment by johnkerl** [comment]
#924 for follow-up doc work
